### PR TITLE
Run CMC metrics

### PR DESCRIPTION
**Pull request by tiffapedia**
Implement a way to run CMC metrics

### CODE REVIEW

**Comment by tiffapedia** [comment]
It can be merged in when I put the resulting graph here

**Comment by agude** [comment]
The first time I run the script I get:

```agude@roosevelt:~/Projects/pelops/experiment_api$ ./run_experiment.sh 
================================================================================
Traceback (most recent call last):
  File "/home/agude/bin/anaconda3/lib/python3.5/random.py", line 253, in choice
    i = self._randbelow(len(seq))
  File "/home/agude/bin/anaconda3/lib/python3.5/random.py", line 230, in _randbelow
    r = getrandbits(k)          # 0 <= r < 2**k
ValueError: number of bits must be greater than zero

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "experiment.py", line 281, in <module>
    main(args)
  File "experiment.py", line 257, in main
    for camset in exp.generate():
  File "experiment.py", line 202, in generate
    output.append(self.__get_camset())
  File "experiment.py", line 175, in __get_camset
    random_car_name = random.choice(list(self.list_of_car_names_per_camera[which_camera_id]))
  File "/home/agude/bin/anaconda3/lib/python3.5/random.py", line 255, in choice
    raise IndexError('Cannot choose from an empty sequence')
IndexError: Cannot choose from an empty sequence
```

The second time (after running `ls` on the directory), it works fine...

**Comment by agude** [comment]
Got the error again after running five or six times. This is a real bug.

**Comment by tiffapedia** [comment]
The bug happens because I was assuming that the data contains camera from 1-20, but the test dataset only contains camera from 1-19.

**Comment by tiffapedia** [comment]
I have fixed the bug. Please test it now and let me know.

File: experiment_api/experiment.py [hunk 1]
```diff
@@ -62,51 +67,33 @@
...
-    num_train_images = 37779
-    total_images = 49358
+from utils import *
 
 
 class ExperimentGenerator(object):
-    def __init__(self, veri_unzipped_path, num_cams, num_cars_per_cam, drop_percentage, seed, time, otherMetric):
-        # set inputs
+    def __init__(self, veri_unzipped_path, num_cams, num_cars_per_cam, drop_percentage, seed, time, typ):
+        # set inputs
```
Comment (agude) [inline experiment_api/experiment.py]: Is your pre-commit hook set up correctly? I think it should strip trailing spaces.

File: experiment_api/experiment.py [hunk 2]
```diff
@@ -128,72 +115,92 @@ def __merge_names(self, *filepaths):
...
         camset = set()
-        # determine whether or not to add the target car
+        which_camera_id = random.choice(self.list_of_cameras)
+        # determine whether or not to add a different target car image
         if not should_drop(self.drop_percentage):
             num_imgs_per_camset = num_imgs_per_camset - 1
-            target_car_img = random.choice(self.list_of_cars[self.target_car])
-            camset.add(target_car_img)
+            while True:
+                # WARNING: If the car is only taken by one camera, this will go into an infinite loop
```
Comment (agude) [inline experiment_api/experiment.py]: That seems like a bad failure mode, can we check for it?

File: experiment_api/experiment.py [hunk 3]
```diff
@@ -128,72 +115,92 @@ def __merge_names(self, *filepaths):
...
         # grab images
-        random_cam = random.choice(list(self.list_of_cameras_per_car[self.target_car]))
         exist_car = list()
-        exist_timestamp = list()
+        #exist_timestamp = list()
         for i in range(0, num_imgs_per_camset):
             while True:
-                random_car = random.choice(list(self.list_of_cars_per_camera[random_cam]))
-                random_car_img = random.choice(self.list_of_cars[random_car])
+                # WARNING: If the dataset only contains one car, this will go into an infinite loop
```
Comment (agude) [inline experiment_api/experiment.py]: Fix.

File: experiment_api/experiment.py [hunk 4]
```diff
@@ -224,35 +231,11 @@ def __eq__(self, other):
...
-        a boolean whether to drop or not drop the image
-    """
-    return random.random() < drop_percentage
-
-# -----------------------------------------------------------------------------
 #  Execution example
 # -----------------------------------------------------------------------------
 
 
+@timewrapper
```
Comment (agude) [inline experiment_api/experiment.py]: Do we want to keep profiling in the mainline?

File: experiment_api/experiment.py [hunk 5]
```diff
@@ -128,72 +115,92 @@ def __merge_names(self, *filepaths):
...
-                    new_timestamp = random_car_img.timestamp
+                    old_timestamp = exist_timestamp[exist_car.index(random_car.car_id)]
+                    new_timestamp = random_car.timestamp
                     valid_timestamp = abs(new_timestamp - old_timestamp) / datetime.timedelta(minutes=1)
                     if valid_timestamp > self.time:
                         break
-            exist_car.append(random_car_img.car_id)
-            exist_timestamp.append(random_car_img.timestamp)
-            camset.add(random_car_img)
+                """
```
Comment (agude) [inline experiment_api/experiment.py]: Weird indenting levels...

File: experiment_api/install.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# python 3
```
Comment (agude) [inline experiment_api/install.sh]: This file looks out of place. We do need a requirements.txt somewhere, but I think an install script is not desired.

File: experiment_api/metric.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+import logging
+import matplotlib.pyplot as plt
+import os
+import random
+import scipy
+import scipy.spatial
+import sys
+
+from experiment import ExperimentGenerator
+from utils import *
```
Comment (agude) [inline experiment_api/metric.py]: `import *` is generally frowned upon; prefer `import utils` and then `utils.foo()` or `from utils import foo(), bar(), baz()`.

File: experiment_api/metric.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+# -----------------------------------------------------------------------------
+#  Execution example
+# -----------------------------------------------------------------------------
+
+def main(args):
+    # extract arguments from command line
+    try:
+        # which metrics
+        if args["--cmc"]:
```
Comment (agude) [inline experiment_api/metric.py]: Does doc ops not handle this?

File: experiment_api/utils.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    train_label_filepath = "train_label.xml"
+    num_cars = 776
+    num_cams = 20
+    num_query_images = 1679
+    num_test_images = 11580
+    num_train_images = 37779
+    total_images = 49358
+
+
+class ImageType(object):
```
Comment (agude) [inline experiment_api/utils.py]: Consider an `enum`: https://docs.python.org/3/library/enum.html

File: experiment_api/experiment.py (hunk 1 above)
Comment (tiffapedia) [inline experiment_api/experiment.py]: I could not fix the pre-commit hooks. When I run pre-commit autoupdate, it prints out the following:

Updating https://github.com/pre-commit/pre-commit...updating 8837cfa7ffcc419216d4e01392cee0f1ceee9c88 -> v0.9.4.
Updating git@github.com:pre-commit/pre-commit-hooks...already up to date.
Updating git@github.com:asottile/reorder_python_imports...already up to date.
Updating git@github.com:Lab41/verboten_words.git...[INFO] Initializing environment for git@github.com:Lab41/verboten_words.git.
An unexpected error has occurred: CalledProcessError: Command: ('/usr/bin/git', 'reset', 'f07c2c0116d0bd6466dde6af4ef9bca5531d6cc7', '--hard')
Return code: 128
Expected return code: 0
Output: (none)
Errors: 
    fatal: Could not parse object 'f07c2c0116d0bd6466dde6af4ef9bca5531d6cc7'.

File: experiment_api/experiment.py (hunk 5 above)
Comment (tiffapedia) [inline experiment_api/experiment.py]: This is an old commit? I am not seeing this.

File: experiment_api/experiment.py (hunk 3 above)
Comment (tiffapedia) [inline experiment_api/experiment.py]: I will try to make that check.

File: experiment_api/experiment.py (hunk 2 above)
Comment (tiffapedia) [inline experiment_api/experiment.py]: I will try to make that check.

File: experiment_api/experiment.py (hunk 4 above)
Comment (tiffapedia) [inline experiment_api/experiment.py]: I'll just comment it for now in case we wanted to profile it again in the future.

**Pull request merged**